### PR TITLE
Feature quantile bfloat16

### LINKED ISSUE #16641: Experimental quantile sketch based on bfloat16 histogram.
`bfloat16` is a 16-bit floating point data type that is the same as the corresponding most significant 16 bits of the `float`.
https://en.wikipedia.org/wiki/Bfloat16_floating-point_format

To calculate quantile, simply convert input value to 16 bit (convert to float, then take the most significant 16 bits), and calculate the histogram of these values.

Hash table is the preferred way to store histogram, because the number of distinct values is small:
```
SELECT uniq(bfloat)
FROM 
(
    SELECT
        number,
        toFloat32(number) AS f,
        bitShiftRight(bitAnd(reinterpretAsUInt32(reinterpretAsFixedString(f)), 4294901760) AS cut, 16),
        reinterpretAsFloat32(reinterpretAsFixedString(cut)) AS bfloat
    FROM numbers(100000000)
)

┌─uniq(bfloat)─┐
│         2623 │
└──────────────┘
```
(when increasing the range of values 1000 times, the number of distinct bfloat16 values increases just by 1280).

Then calculate quantile from the histogram.

This sketch is very simple and rough. Its relative precision is constant 1 / 256 = 0.390625%.

Linear array of values can be used if the number of values is very small, similar to `QuantileTimingTiny`.

### PR DESCRIPTION
**Pull request by RedClusive**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `quantileBFloat16` aggregate function as well as the corresponding `quantilesBFloat16` and `medianBFloat16`. It is very simple and fast quantile estimator with relative error not more than 0.390625%. This closes #16641.

Detailed description / Documentation draft:

Made new aggregate function that computes an approximate quantile of a numeric data sequence.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Let's rename `quantileBfloat16` -> `quantileBFloat16`.

**Comment by alexey-milovidov** [comment]
AST Fuzzer: https://github.com/ClickHouse/ClickHouse/issues/23714

File: src/AggregateFunctions/QuantileBFloat16Histogram.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+    using Histogram = BFloat16Histogram<Value>;
+    Histogram data;
+
+    void add(const Value & x) { data.add(x); }
+
+    template <typename Weight>
+    void add(const Value &, const Weight &)
+    {
+        throw Exception("Method add with weight is not implemented for QuantileBFloat16Histogram", ErrorCodes::NOT_IMPLEMENTED);
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/QuantileBFloat16Histogram.h:23]: Why not? Simply add the weight.

File: src/AggregateFunctions/QuantileBFloat16Histogram.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NOT_IMPLEMENTED;
+}
+
+template <typename Value>
+struct QuantileBFloat16Histogram
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/QuantileBFloat16Histogram.h:13]: It's pointless to have this small wrapper class. We can implement BFloat16Histogram directly here.

File: src/AggregateFunctions/BFloat16Histogram.h [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+#include <ext/bit_cast.h>
+
+namespace DB
+{
+
+template <typename Value>
+struct BFloat16Histogram
+{
+    using bfloat16 = UInt16;
+    using Data = HashMap<bfloat16, size_t>;
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/BFloat16Histogram.h:18]: I don't understand, why we use *both* hash table and array?

Comment (alexey-milovidov) [reply src/AggregateFunctions/BFloat16Histogram.h:18]: I thought hash table is enough.

Comment (alexey-milovidov) [reply src/AggregateFunctions/BFloat16Histogram.h:18]: > Linear array of values can be used if the number of values is very small, similar to QuantileTimingTiny.

Comment (alexey-milovidov) [reply src/AggregateFunctions/BFloat16Histogram.h:18]: See also quantileExactWeighted.

Comment (RedClusive) [reply src/AggregateFunctions/BFloat16Histogram.h:18]: Array used to avoid sorting array every time 'get' called. Is it useless?

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Alright.

> Performance — 2 errors, 42 unstable 

The function does not exist on old server.

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-9779.

**Comment by Azmisov** [comment]
Can you explain how you derived the relative precision being 1/256, 0.390625%?

**Comment by Azmisov** [comment]
I believe it is incorrect. First of all, maximum error for float64 and greater is **infinite**, e.g. when they overflow when converting to float32. I think its important to clarify and warn people this will not be a good method for numbers outside of float32 exponent range.

But disregarding that, let's just consider float32:

- `relative_error = (estimate - value)/value = estimate/value - 1`
- `estimate` will be the same as `value`, but with truncated significand
- `rel_err = (sign*exp*(1+frac_hi))/(sign*exp*(1 + frac_lo + frac_hi)) - 1 = (1 + frac_hi)/(1 + frac_hi + frac_lo) - 1 = -frac_lo/(1 + frac_hi + frac_lo) `
- `rel_err <= 0`
- maximum absolute rel err is when `frac_lo` is large and `frac_hi` is small
- `max(frac_lo) = sum{i=8 to 23}(1/2^i) = 0.007812380790710449`
- `min(frac_hi) = 0`

Which gives a float32 relative error range of **[-0.7751820616235141%, 0%]**

For float64/128/etc, assuming the value is representable as float32, error should be pretty close. Though they will pick the next representable float32, not simply truncating, so a little tricky to get an exact value.

**Pull request merged**